zenodo/modules/records/serializers/schemaorg.py

```diff
@@ -0,0 +1,85 @@
+# useful, but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+# General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with Zenodo; if not, write to the
+# Free Software Foundation, Inc., 59 Temple Place, Suite 330, Boston,
+# MA 02111-1307, USA.
+#
+# In applying this license, CERN does not
+# waive the privileges and immunities granted to it by virtue of its status
+# as an Intergovernmental Organization or submit itself to any jurisdiction.
+
+"""Zenodo schema.org serializer."""
+
+from __future__ import absolute_import, print_function
+
+from zenodo.modules.records.models import ObjectType
+from zenodo.modules.records.serializers.schemas import schemaorg as schemas
+
+from .json import ZenodoJSONSerializer
+
+
+class ZenodoSchemaOrgSerializer(ZenodoJSONSerializer):
+    """Zenodo schema.org serializer.
+
+    Serializes the record using the appropriate marshmallow schema based on
+    its schema.org type.
+    """
+
+    def funcname(parameter_list):
+        pass
+
+    SCHEMA_ORG_TYPES = {
+        'publication': schemas.ScholarlyArticle,
+        'poster': schemas.CreativeWork,
+        'presentation': schemas.PresentationDigitalDocument,
+        'dataset': schemas.Dataset,
+        'image': schemas.ImageObject,
+        'video': schemas.MediaObject,
+        'software': schemas.SoftwareSourceCode,
+        'lesson': schemas.CreativeWork,
+        'other': schemas.CreativeWork,
+        'publication-book': schemas.Book,
+        'publication-section': schemas.ScholarlyArticle,
+        'publication-conferencepaper': schemas.ScholarlyArticle,
+        'publication-article': schemas.ScholarlyArticle,
+        'publication-patent': schemas.CreativeWork,
+        'publication-preprint': schemas.ScholarlyArticle,
+        'publication-report': schemas.ScholarlyArticle,
+        'publication-softwaredocumentation': schemas.CreativeWork,
+        'publication-thesis': schemas.ScholarlyArticle,
+        'publication-technicalnote': schemas.ScholarlyArticle,
+        'publication-workingpaper': schemas.ScholarlyArticle,
+        'publication-proposal': schemas.CreativeWork,
+        'publication-deliverable': schemas.CreativeWork,
+        'publication-milestone': schemas.CreativeWork,
+        'publication-other': schemas.CreativeWork,
+        'image-figure': schemas.ImageObject,
+        'image-plot': schemas.ImageObject,
+        'image-drawing': schemas.ImageObject,
+        'image-diagram': schemas.ImageObject,
+        'image-photo': schemas.Photograph,
+        'image-other': schemas.ImageObject,
+    }
+
+    def dump(self, obj, context=None):
+        """Serialize object with schema."""
+        data = obj['metadata']
+        obj_type = ObjectType.get_by_dict(data['resource_type'])
+        internal_id = obj_type['internal_id']
+        schema_cls = self.SCHEMA_ORG_TYPES.get(
```

Comment:
maybe we could skip the double config (here and objecttypes.json) and resolve this with something like ``schemas.getattr(obj_type['schema.org'].split('https://schema.org/')[1])``?

Reply:
The only problem I see with that is that if somebody changes some `schema.org` key from `objecttypes.json`, the app will still be runnable. If we have tests for all `internal_id -> schema.org` types then it's not an issue of course.

Reply:
1) I would do the mapping based on ``obj_type['schema.org']`` instead of ``obj_type['internal_id']``. That should at least make the mapping somewhat smaller.
2) However, probably something like @krzysztof suggests is as good:
``schema_cls = getattr(schemas, obj_type['schema.org'].split('https://schema.org/')[1], schemas.CreativeWork)``